doc/developer/design/19042023_stash_migrations.md

```diff
@@ -0,0 +1,352 @@
+  * Fixed previous migrations that failed to remove old types from the Stash.
+* `_incident-47`
+  * Scaling issue, the SQL statement we were running on CockroachDB was >32MiB which was the allowed size.
+
+
+# Explanation
+[explanation]: #explanation
+
+### What is the "Stash"?
+
+The [Stash](https://github.com/MaterializeInc/materialize/blob/23d7cad1211b4d1d1de3a26652937f9f2c4df61c/src/stash/src/lib.rs#L107-L124) 
+is a time-varying key-value store, i.e. it stores colletions of `(key, value, timestamp, diff)`, 
+which we use to persist metadata for a Materialize deployment. For example, we use the Stash to
+persist metadata for all `MATERIALIZED VIEWS` that a user has created. This way when restarting an 
+environment, we know what Materialized Views to recreate on Compute. Concretely we use CockroachDB
+and all keys and values are serialized as JSON for human readability.
+
+### Part 1: Snapshotting Types
+
+What makes Stash migrations hard to write and hard to reason about, is that you need to define a
+type that supports deserializing the current Stash and serializing the new Stash. For example, if
+you want to add a field to a struct you need to wrap the new field in an `Option<...>`. This way
+your struct can deserialize when the field doesn't exist, yet provide the Stash with the new data.
+There currently is not a way to do a migration like, converting a type from a `u64` to an `enum`.
+
+The proposal to fix this is to "snapshot" our types, this would allow us to, in a type safe way,
+represent the current data we're reading from, and the new types we want to migrate to. Then
+we'd be able to do arbitrarily complex migrations from one type to another.
+
+Concretely what this means is having types like:
+
+```
+struct RoleValuesV5 {
+    create_db: bool,
+}
+
+struct RoleValuesV6 {
+    // A bit flag of permissions this user is allowed.
+    perms: u64,
+}
+```
+
+### Part 2: Structuring Migrations
+
+Today migrations are a [list of closures](https://github.com/MaterializeInc/materialize/blob/23d7cad1211b4d1d1de3a26652937f9f2c4df61c/src/adapter/src/catalog/storage.rs#L78), and we index
+into this list based off of the current version we read from our database. This is fragile
+because the list always need to be ever increasing in length, we can't move the position of a
+migration in the list, and there is nothing to indicate that a new migration needs to be added.
+
+The proposal is to maintain a notion of a "current stash version" and restructure the migrations
+based on this version, for example:
+
+```
+const STASH_VERSION: u64 = 3;
+
+async fn migrate(...) -> Result<...> {
+    let version = stash.read_version().await?;
+
+    const NEXT_VERSION = STASH_VERSION + 1;
+    match version {
+        0 => // migrate
+        1 => // migrate
+        2 => // migrate
+        3 => // migrate
+        NEXT_VERSION.. => panic!("Traveled to the future!"),
+    }
+}
```

Comment:
Currently, the first migration is used to set up the stash with initial values.

https://github.com/MaterializeInc/materialize/blob/b38572937e354e93de330a2a36ee76ab676edf3e/src/adapter/src/catalog/storage.rs#L86-L424

Where does that initialization step fit into this proposal?

Reply:
Added this to the design doc as it's own section, tl;dr I propose we add a new "initialize" step to the Stash which populates the Stash with the necessary data, and sets the Stash version to the `STASH_VERSION` we have in code.